sdk/main/src/impl/NodeAddress.cc

```diff
@@ -26,42 +28,159 @@
+  std::shared_ptr<Endpoint> endpoint = std::make_shared<Endpoint>(ipAddress, port);
+  mEndpoints.push_back(endpoint);
+}
+
+//-----
 NodeAddress NodeAddress::fromProtobuf(const proto::NodeAddress& protoNodeAddress)
 {
   NodeAddress outputNodeAddress;
 
   for (int i = 0; i < protoNodeAddress.serviceendpoint_size(); ++i)
   {
-    outputNodeAddress.mEndpoints.push_back(internal::Endpoint::fromProtobuf(protoNodeAddress.serviceendpoint(i)));
+    Endpoint endpoint = Endpoint::fromProtobuf(protoNodeAddress.serviceendpoint(i));
+    outputNodeAddress.mEndpoints.push_back(std::make_shared<Endpoint>(endpoint));
   }
 
   if (protoNodeAddress.ipaddress().length() != 0)
   {
-    outputNodeAddress.mEndpoints.emplace_back(internal::IPv4Address::fromString(protoNodeAddress.ipaddress()),
-                                              protoNodeAddress.portno());
+    Endpoint endpoint = Endpoint(IPv4Address::fromString(protoNodeAddress.ipaddress()), protoNodeAddress.portno());
+    outputNodeAddress.mEndpoints.emplace_back(std::make_shared<Endpoint>(endpoint));
   }
 
   outputNodeAddress.mRSAPublicKey = protoNodeAddress.rsa_pubkey();
   outputNodeAddress.mNodeId = protoNodeAddress.nodeid();
-  outputNodeAddress.mCertificateHash = protoNodeAddress.nodecerthash();
+  outputNodeAddress.mNodeAccountId = AccountId::fromProtobuf(protoNodeAddress.nodeaccountid());
+  outputNodeAddress.mNodeCertHash = protoNodeAddress.nodecerthash();
   outputNodeAddress.mDescription = protoNodeAddress.description();
-  outputNodeAddress.mAccountId = AccountId::fromProtobuf(protoNodeAddress.nodeaccountid());
+  outputNodeAddress.mStake = protoNodeAddress.stake();
 
   return outputNodeAddress;
 }
 
+//-----
+NodeAddress NodeAddress::fromString(const std::string& nodeAddress)
+{
+  std::vector<std::string> parts;
+  std::stringstream strStream(nodeAddress);
+  std::string ipAddressV4;
+  int port;
+
+  try
+  {
+    std::string temp;
+
+    while (getline(strStream, temp, ':'))
+    {
+      parts.push_back(temp.c_str());
+    }
+
+    ipAddressV4 = parts[0].c_str();
+    port = atoi(parts[1].c_str());
+  }
+  catch (const std::exception& e)
+  {
+    throw IllegalStateException("Failed to parse the node address.");
+  }
+
+  return NodeAddress(ipAddressV4, port);
+}
+
+//-----
+NodeAddress& NodeAddress::setRSAPublicKey(const std::string& publicKey)
+{
+  mRSAPublicKey = publicKey;
+  return *this;
+}
+
+//-----
+NodeAddress& NodeAddress::setNodeId(const int64_t nodeId)
+{
+  mNodeId = nodeId;
+  return *this;
+}
+
+//-----
+NodeAddress& NodeAddress::setNodeAccountId(const AccountId& accountId)
+{
+  mNodeAccountId = accountId;
+  return *this;
+}
+
+//-----
+NodeAddress& NodeAddress::setNodeCertHash(const std::string& certHash)
+{
+  mNodeCertHash = certHash;
+  return *this;
+}
+
+//-----
+NodeAddress& NodeAddress::setEndpoints(const std::vector<std::shared_ptr<Endpoint>>& endpoints)
+{
+  mEndpoints.clear();
+  mEndpoints.assign(endpoints.begin(), endpoints.end());
```

Comment:
Could this just be `mEndpoints = endpoints;`?

Reply:
Updated in c990434b4fffd19b4aadb7141bfe82ae492c129d.